pkg/pillar/cmd/nodeagent/handletimers.go

```diff
@@ -0,0 +1,188 @@
+	// partition is not in inprogress state,
+	// nothing to do
+	if !ctx.updateInprogress || ctx.testInprogress {
+		return
+	}
+	// only when current partition state is in progress
+	// start the test
+	ctx.upgradeTestStartTime = time.Now()
+	ctx.testInprogress = true
+	successLimit := time.Second *
+		time.Duration(globalConfig.MintimeUpdateSuccess)
+	ctx.remainingTestTime = successLimit
+	ctx.testInprogress = true
+}
+
+func resetTestStartTime(ctx *nodeagentContext) {
+	if !ctx.testInprogress {
+		return
+	}
+	// the led status has changed, incase, the test is in progress,
+	// we need to start the test again, reset first
+	ctx.testInprogress = false
+	ctx.remainingTestTime = 0
+	ctx.testInprogress = false
+}
+
+func setConfigGetFailState(ctx *nodeagentContext) {
+	if !ctx.configGetFail {
+		ctx.configGetFail = true
+	}
+}
+
+func resetConfigGetFailState(ctx *nodeagentContext) {
+	if ctx.configGetFail {
+		ctx.configGetFail = false
+		ctx.configGetFailCount = 0
+	}
+}
+
+func executeConfigGetFailState(ctx *nodeagentContext) {
+	if !ctx.configGetFail {
+		return
+	}
+	if ctx.configGetFailCount >= maxConfigGetFailCount {
+		resetConfigGetFailState(ctx)
+		resetTestStartTime(ctx)
+	}
+}
+
+func checkUpgradeValidationTestTimeExpiry(ctx *nodeagentContext) bool {
+	if !ctx.testInprogress {
+		return false
+	}
+	timePassed := time.Since(ctx.upgradeTestStartTime)
+	successLimit := time.Second *
+		time.Duration(globalConfig.MintimeUpdateSuccess)
+	if timePassed < successLimit {
+		ctx.remainingTestTime = successLimit - timePassed
+		log.Infof("CurPart: %s inprogress, waiting for %d seconds\n",
+			ctx.curPart, ctx.remainingTestTime/time.Second)
+		publishNodeAgentStatus(ctx)
+		return false
+	}
+	return true
+}
+
+func handleConfigGetFail(ctx *nodeagentContext) {
+	if !ctx.testInprogress {
+		return
+	}
+	if !ctx.configGetFail {
+		ctx.configGetFail = true
+	}
+	ctx.configGetFailCount++
+	executeConfigGetFailState(ctx)
+}
+
+func updateLastConfigReceivedTime(ctx *nodeagentContext,
+	status types.ZedAgentStatus) {
+	log.Debugf("lastConfigReceivedtime: %v, failCount: %d \n",
+		status.LastConfigReceivedTime, status.ConfigGetFailCount)
+	// time stamp has not changed, since last config receive
+	if ctx.lastConfigReceivedTime == status.LastConfigReceivedTime {
+		handleConfigGetFail(ctx)
+		return
+	}
+	// config get is successful
+	resetConfigGetFailState(ctx)
+	setTestStartTime(ctx)
+	ctx.lastConfigReceivedTime = status.LastConfigReceivedTime
+	if checkUpgradeValidationTestTimeExpiry(ctx) {
+		log.Infof("CurPart: %s, Upgrade Validation Test Complete\n",
+			ctx.curPart)
+		resetTestStartTime(ctx)
+		initiateBaseOsZedCloudTestComplete(ctx)
+		publishNodeAgentStatus(ctx)
+	}
+}
+
+func execReboot(ctx *nodeagentContext, rebootStr string) {
+
+	log.Infof("execReboot(): Reboot reason(%s)\n", rebootStr)
+	if ctx.needsReboot {
+		log.Infof("reboot flag is already set\n")
+		return
+	}
+
+	ctx.needsReboot = true
+	ctx.rebootReason = rebootStr
```

Comment:
Where do you write the reboot reason to disk?
zedagent currently calls
			agentlog.RebootReason(errStr)
and you need the same AFAICT

Reply:
While publishing the nodeagent status, it is picked up by the zedagent, nodeagent status handler and pushed as part of device reboot string.

Reply:
Yes, in case zedagent is running. But nodeagent also handles the case when  only client is running, and I assumed that is why you have code where you have nodeagent do the reboot itself. In that case it needs to write the reason.

Reply:
Sure, have added them